### PR TITLE
Revise Router module—>Protocol module  and create new module "api"

### PR DESCRIPTION
**Pull request by a8901053**
Revise Router module—>Protocol module.
Use single source of truth to make sure serialization and deserialization.

### CODE REVIEW

File: .drone.yml [hunk 1]
```diff
@@ -2,8 +2,8 @@ kind: pipeline
 name: default
 
 steps:
-- name: test
-  image: maven:3-jdk-8
-  commands:
-  - mvn test -B
+  - name: test
```
Comment (Johnny850807) [inline .drone.yml:5]: Rollback this.

File: client/src/main/java/tw/waterball/survive/io/client/controller/PlayerController.java [hunk 2]
```diff
@@ -44,7 +44,7 @@ public void playerThrowCurrentWeapon() {
 
     }
 
-    public void playerSwitchWeapon(int weaponIndex) {
+    public void playerSwitchWeapon(int weaponSlotIndex) {
```
Comment (Johnny850807) [inline client/src/main/java/tw/waterball/survive/io/client/controller/PlayerController.java:47]: Great to enhance the name! 👍

File: protocol/src/main/java/tw/waterball/survive/io/router/protocol/api/ClientApi.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
+package tw.waterball.survive.io.router.protocol.api;
+
+
+public interface ClientApi extends Api {
+
+    void onLoginSuccessfully(byte[] map, byte[] player);
```
Comment (Johnny850807) [inline protocol/src/main/java/tw/waterball/survive/io/router/protocol/api/ClientApi.java:6]: This is higher level interface, why `byte` is used here?

File: protocol/src/main/java/tw/waterball/survive/io/router/protocol/api/ClientApi.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+    void onWeaponUsed(int roleId);
+
+    void onPickUpWeapon(int roleId, int weaponId);
+
+    void onTurn(int roleId, float angle);
+
+    void onDead(int id);
+
+    void onWorldObjectJoinsPlayerViewPort(byte[] worldObject);
```
Comment (Johnny850807) [inline protocol/src/main/java/tw/waterball/survive/io/router/protocol/api/ClientApi.java:34]: ditto

File: protocol/src/main/java/tw/waterball/survive/io/router/protocol/factory/ProtocolFactoryImpl.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+import tw.waterball.survive.io.router.protocol.api.ServerApi;
+import tw.waterball.survive.io.router.protocol.exception.ApiClassNotFoundException;
+import tw.waterball.survive.io.router.protocol.protocols.ClientProtocol;
+import tw.waterball.survive.io.router.protocol.protocols.Protocol;
+import tw.waterball.survive.io.router.protocol.protocols.ServerProtocol;
+
+public class ProtocolFactoryImpl implements ProtocolFactory {
+
+    @Override
+    public Protocol createProtocol(Class<? extends Api> apiClass) {
```
Comment (Johnny850807) [inline protocol/src/main/java/tw/waterball/survive/io/router/protocol/factory/ProtocolFactoryImpl.java:14]: No, this is not what I meant. Why didn't you ask?

File: protocol/src/main/java/tw/waterball/survive/io/router/protocol/protocols/ClientProtocol.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
+package tw.waterball.survive.io.router.protocol.protocols;
+
+import tw.waterball.survive.io.router.protocol.api.Api;
+import tw.waterball.survive.io.router.protocol.api.ClientApi;
+import tw.waterball.survive.io.router.protocol.router.Router;
+
+import java.io.InputStream;
+import java.io.OutputStream;
+
+public class ClientProtocol implements Protocol {
```
Comment (Johnny850807) [inline protocol/src/main/java/tw/waterball/survive/io/router/protocol/protocols/ClientProtocol.java:10]: Remove this class

File: protocol/src/main/java/tw/waterball/survive/io/router/protocol/protocols/ServerProtocol.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
+package tw.waterball.survive.io.router.protocol.protocols;
+
+import tw.waterball.survive.io.router.protocol.api.Api;
+import tw.waterball.survive.io.router.protocol.api.ServerApi;
+import tw.waterball.survive.io.router.protocol.router.Router;
+
+import java.io.InputStream;
+import java.io.OutputStream;
+
+public class ServerProtocol implements Protocol {
```
Comment (Johnny850807) [inline protocol/src/main/java/tw/waterball/survive/io/router/protocol/protocols/ServerProtocol.java:10]: Remove this class

File: server/src/main/java/tw/waterball/survive/io/server/worldobject/role/Player.java [hunk 8]
```diff
@@ -1,89 +1,16 @@
...
-    @Override
-    public void onMoveLeftward(Role role) {
-        client.onMoveLeftward(role);
-    }
-
-    @Override
-    public void onMoveRightward(Role role) {
-        client.onMoveRightward(role);
-    }
+public class Player extends Role {
```
Comment (Johnny850807) [inline server/src/main/java/tw/waterball/survive/io/server/worldobject/role/Player.java:6]: `Player` should implement `ClientAPI`

**Review by Wally5077 — APPROVED** [review]
好哦

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
Review done, there are many things go wrong, please ask before you begin, again.

File: protocol/src/main/java/tw/waterball/survive/io/router/protocol/factory/ProtocolFactoryImpl.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+        if (ServerApi.class.equals(apiClass)) {
+            return new ServerProtocol();
+        } else if (ClientApi.class.equals(apiClass)) {
+            return new ClientProtocol();
+        }
+        throw new ApiClassNotFoundException();
```
Comment (Johnny850807) [inline protocol/src/main/java/tw/waterball/survive/io/router/protocol/factory/ProtocolFactoryImpl.java:20]: No, this is not what I meant. Why didn't you ask?

**Review by Johnny850807 — CHANGES_REQUESTED** [review]

File: protocol/src/main/java/tw/waterball/survive/io/router/protocol/factory/ProtocolFactoryImpl.java (hunk 9 above)
Comment (Johnny850807) [reply protocol/src/main/java/tw/waterball/survive/io/router/protocol/factory/ProtocolFactoryImpl.java:20]: If you use reflection, why do you need to couple the `ProtocolFactoryImpl` with the subclass of `Api`?

File: protocol/src/main/java/tw/waterball/survive/io/router/protocol/api/ClientApi.java (hunk 3 above)
Comment (a8901053) [reply protocol/src/main/java/tw/waterball/survive/io/router/protocol/api/ClientApi.java:6]: 如果要用WorldObject去傳的話 要加入server moudel的依賴 這樣就變成循環依賴

File: client/src/main/java/tw/waterball/survive/io/client/controller/EventController.java [hunk 10]
```diff
@@ -1,53 +1,88 @@
...
+import java.util.List;
 
+public class EventController implements ClientApi {
     private SpriteContainer spriteContainer;
     private SpritePrototypeFactory spritePrototypeFactory;
     private GameView gameView;
 
-    public void onSufferingInjury(int id, int damage) {
+    @Override
+    public void onLoginSuccessfully(List<Object> map, List<Object> playersInViewPort, Object player) {
```
Comment (Johnny850807) [inline client/src/main/java/tw/waterball/survive/io/client/controller/EventController.java:15]: Do not use Object, use DTOs.

File: client/src/main/java/tw/waterball/survive/io/client/controller/EventController.java [hunk 11]
```diff
@@ -1,53 +1,88 @@
...
     }
 
+    @Override
     public void onDead(int id) {
 
     }
 
-    public void onSpriteJoinsPlayerViewPort(Sprite sprite) {
+    @Override
+    public void onWorldObjectJoinsPlayerViewPort(Object worldObject) {
```
Comment (Johnny850807) [inline client/src/main/java/tw/waterball/survive/io/client/controller/EventController.java:85]: ditto

File: protocol/src/main/java/tw/waterball/survive/io/router/protocol/api/Api.java [hunk 12]
```diff
@@ -0,0 +1,4 @@
+package tw.waterball.survive.io.router.protocol.api;
+
+public interface Api {
```
Comment (Johnny850807) [inline protocol/src/main/java/tw/waterball/survive/io/router/protocol/api/Api.java:3]: We needn't this ;)

File: protocol/src/main/java/tw/waterball/survive/io/router/protocol/api/ClientApi.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
+package tw.waterball.survive.io.router.protocol.api;
+
+
+import java.util.List;
+
+public interface ClientApi extends Api {
+
+    void onLoginSuccessfully(List<Object> map, List<Object> playersInViewPort, Object player);
```
Comment (Johnny850807) [inline protocol/src/main/java/tw/waterball/survive/io/router/protocol/api/ClientApi.java:8]: ditto

File: protocol/src/main/java/tw/waterball/survive/io/router/protocol/api/ClientApi.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+
+    void onWeaponUsed(int roleId);
+
+    void onPickUpWeapon(int roleId, int weaponId);
+
+    void onTurn(int roleId, float angle);
+
+    void onDead(int id);
+
+    void onWorldObjectJoinsPlayerViewPort(Object worldObject);
```
Comment (Johnny850807) [inline protocol/src/main/java/tw/waterball/survive/io/router/protocol/api/ClientApi.java:36]: ditto

File: protocol/src/main/java/tw/waterball/survive/io/router/protocol/protocols/StandardProtocol.java [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+            this.in = in;
+        }
+
+        @Override
+        public void startRouting() {
+
+        }
+    }
+
+    private static class StandardClientApi implements ClientApi {
```
Comment (Johnny850807) [inline protocol/src/main/java/tw/waterball/survive/io/router/protocol/protocols/StandardProtocol.java:50]: We needn't this. We have reflection.

File: protocol/src/main/java/tw/waterball/survive/io/router/protocol/protocols/StandardProtocol.java [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+
+        }
+
+        @Override
+        public void onWorldObjectJoinsPlayerViewPort(Object worldObject) {
+
+        }
+    }
+
+    private static class StandardServerApi implements ServerApi {
```
Comment (Johnny850807) [inline protocol/src/main/java/tw/waterball/survive/io/router/protocol/protocols/StandardProtocol.java:133]: We needn't this. We have reflection

File: server/src/main/java/tw/waterball/survive/io/server/worldobject/role/Player.java [hunk 17]
```diff
@@ -1,89 +1,93 @@
...
+    }
 
     public void reload() {
 
     }
 
     @Override
-    public void onLoginSuccessfully() {
-        client.onLoginSuccessfully();
+    public void onLoginSuccessfully(List<Object> map, List<Object> playersInViewPort, Object player) {
```
Comment (Johnny850807) [inline server/src/main/java/tw/waterball/survive/io/server/worldobject/role/Player.java:20]: ditto

File: server/src/main/java/tw/waterball/survive/io/server/worldobject/role/Player.java [hunk 18]
```diff
@@ -1,89 +1,93 @@
...
-    public void onDead(Role role) {
-        client.onDead(role);
+    public void onDead(int id) {
+        clientApi.onDead(id);
     }
 
     @Override
-    public void onWorldObjectJoinsPlayerViewPort(WorldObject worldObject) {
-        client.onWorldObjectJoinsPlayerViewPort(worldObject);
+    public void onWorldObjectJoinsPlayerViewPort(Object worldObject) {
```
Comment (Johnny850807) [inline server/src/main/java/tw/waterball/survive/io/server/worldobject/role/Player.java:90]: ditto

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
Review done, great.

File: api/src/main/java/tw/waterball/survive/io/api/api/ClientApi.java [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+import tw.waterball.survive.io.api.dto.RoleDTO;
+import tw.waterball.survive.io.api.dto.TerrainDTO;
+import tw.waterball.survive.io.api.dto.event.*;
+
+import java.util.List;
+
+public interface ClientApi {
+
+    void onLoginSuccessfully(List<TerrainDTO> terrainDTOs,
+                             List<RoleDTO> roleDTOs, RoleDTO roleDTO);
```
Comment (Johnny850807) [inline api/src/main/java/tw/waterball/survive/io/api/api/ClientApi.java:13]: You have to distinguish the second parameter and the third one.

File: api/src/main/java/tw/waterball/survive/io/api/dto/GunDTO.java [hunk 20]
```diff
@@ -0,0 +1,9 @@
+package tw.waterball.survive.io.api.dto;
+
+import lombok.Data;
+
+@Data
+public class GunDTO {
+    private int bulletsShotAtOnce;
+    private MagazineDTO magazine;
```
Comment (Johnny850807) [inline api/src/main/java/tw/waterball/survive/io/api/dto/GunDTO.java:8]: `magazineDTO`

File: api/src/main/java/tw/waterball/survive/io/api/dto/RoleDTO.java [hunk 21]
```diff
@@ -0,0 +1,18 @@
...
+
+@Data
+public class RoleDTO {
+    private String name;
+    private int money;
+    private int speed;
+    private int def;
+    private int hp;
+    private int maxHp;
+    private List<WeaponSlotDTO> weaponSlots;
```
Comment (Johnny850807) [inline api/src/main/java/tw/waterball/survive/io/api/dto/RoleDTO.java:15]: `weaponSlotDTOs`

File: api/src/main/java/tw/waterball/survive/io/api/dto/RoleDTO.java [hunk 22]
```diff
@@ -0,0 +1,18 @@
...
+    private List<MagazineDTO> magazines;
```
Comment (Johnny850807) [inline api/src/main/java/tw/waterball/survive/io/api/dto/RoleDTO.java:16]: `magazineDTOs`

File: api/src/main/java/tw/waterball/survive/io/api/dto/RoleDTO.java [hunk 23]
```diff
@@ -0,0 +1,18 @@
...
+    private WeaponSlotDTO currentWeaponSlot;
```
Comment (Johnny850807) [inline api/src/main/java/tw/waterball/survive/io/api/dto/RoleDTO.java:17]: `currentWeaponSlotDTO`

File: client/src/main/java/tw/waterball/survive/io/client/controller/EventController.java [hunk 24]
```diff
@@ -1,53 +1,87 @@
...
 
+public class EventController implements ClientApi {
     private SpriteContainer spriteContainer;
     private SpritePrototypeFactory spritePrototypeFactory;
     private GameView gameView;
 
-    public void onSufferingInjury(int id, int damage) {
+    @Override
+    public void onLoginSuccessfully(List<TerrainDTO> terrainDTOs, List<RoleDTO> roleDTOs,
+                                    RoleDTO roleDTO) {
```
Comment (Johnny850807) [inline client/src/main/java/tw/waterball/survive/io/client/controller/EventController.java:19]: ditto

File: protocol/src/main/java/tw/waterball/survive/io/protocol/exception/ApiClassNotFoundException.java [hunk 25]
```diff
@@ -0,0 +1,22 @@
+package tw.waterball.survive.io.protocol.exception;
+
+public class ApiClassNotFoundException extends RuntimeException {
```
Comment (Johnny850807) [inline protocol/src/main/java/tw/waterball/survive/io/protocol/exception/ApiClassNotFoundException.java:3]: Do you need this class? Remove it.

File: server/src/main/java/tw/waterball/survive/io/server/worldobject/role/Player.java [hunk 26]
```diff
@@ -1,89 +1,93 @@
...
 
     public void reload() {
 
     }
 
     @Override
-    public void onLoginSuccessfully() {
-        client.onLoginSuccessfully();
+    public void onLoginSuccessfully(List<TerrainDTO> terrainDTOs, List<RoleDTO> roleDTOs,
+                                    RoleDTO roleDTO) {
```
Comment (Johnny850807) [inline server/src/main/java/tw/waterball/survive/io/server/worldobject/role/Player.java:24]: ditto

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
Review done

**Review by timt1028 — APPROVED** [review]
LGTM

File: server/src/main/java/tw/waterball/survive/io/server/worldobject/role/Player.java [hunk 27]
```diff
@@ -1,89 +1,97 @@
...
+    }
 
     public void reload() {
 
     }
 
     @Override
-    public void onLoginSuccessfully() {
-        client.onLoginSuccessfully();
+    public void onLoginSuccessfully(RoleDTO roleDTO) {
```
Comment (Johnny850807) [inline server/src/main/java/tw/waterball/survive/io/server/worldobject/role/Player.java:23]: `loginPlayerDTO`, this is why you have to use ctrl+F6

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
1 minor issue

File: server/src/main/java/tw/waterball/survive/io/server/worldobject/role/Player.java (hunk 27 above)
Comment (a8901053) [reply server/src/main/java/tw/waterball/survive/io/server/worldobject/role/Player.java]: Ok thanks

**Review by Johnny850807 — APPROVED** [review]
LGTM

**Pull request merged**